src/protocol/securemul.rs

```diff
@@ -14,6 +16,53 @@ pub struct DValue<F> {
     d: F,
 }
 
+#[derive(Debug)]
+pub struct SecureMulAll<'a, N> {
+    prss: &'a PrssSpace,
+    gateway: &'a Gateway<N>,
+    step: &'a UniqueStepId,
+}
+
+impl<'a, N: Network> SecureMulAll<'a, N> {
+    pub fn new(prss: &'a PrssSpace, gateway: &'a Gateway<N>, step: &'a UniqueStepId) -> Self {
```

Comment:
Is there ever a case where you are going to execute this multiple times in parts?  I want to ensure that starting from 0 each time is going to work properly.

If it doesn't, then this could take a starting index.

Reply:
For now, I think we are okay with starting from 0 each time.

Just out of curiosity, if we were to fix this, instead of taking a starting index, could we `narrow()` the context? Running this under a sub-context is unintended for the caller, but does it result in a bad state?

Reply:
I think using the index part is okay, because the caller also uses the index from `for i in ...` alike to generate record IDs. 

The part I'm not sure of is whether there is a case where we execute this multiple times in parts, as you and Martin said. Would there be a case where we do something like this?

```
let v1, v2: Vec<Replicated<F>> = ...
multiply_all(v1[..x], v2[..x]);
...
multiply_all(v1[x+1..], v2[x+1..]);
```

I think narrow() would cover this case nicely. That doesn't force the caller to keep track and pass a start index each time.